feat: add ability to specify a base item whose values are used for the named item before randomizing

LordMidas opened the pull request:
We should make a list of fields in `::MSU.Items.BaseWeaponFields` and iterate over that to allow easy way for modders to change the fields which are copied over from the Base Weapon. I will do this after #176 is merged (because of folder structure, to avoid resolving conflicts).

LordMidas committed:
feat: add ability to specify a base weapon whose values are used for the named weapon before randomizing

TaroEld reviewed (changes requested):
Put this in a new function, like setBaseWeaponValues or something that is then called before randomizeValues() so that it can be amended/overwritten for weapons with special stats or something. Good idea

LordMidas committed:
refactor: use a separate function to set the base values

TaroEld commented:
postponed for 1.3 with potential changes

LordMidas committed:
refactor: add a separate function to get the value and rename the functions

TaroEld reviewed (approved):
Approved, but for parity we should do the same for the other named item categories.

LordMidas committed:
refactor: rename to BaseItemScript

LordMidas committed:
feat: add BaseItemScript to named_shield

LordMidas committed:
feat: add BaseItemScript to named_armor

LordMidas committed:
feat: add BaseItemScript to named_helmet

LordMidas commented:
I'm still not 100% happy with this. If a modder wants to add more fields that should be copied from the base item, the only way to do that right now would be to hook the `getValuesForRandomize` function, instantiate the base item, add the new values to the return of the hooked function, and then return that. 

We should instead try to have a function that returns an array of strings which are the fields that should be copied from the base item. Then modders can hook that function and modify the list - and the instantiation of the base item only happens once within the msu function that copies the values.

LordMidas commented:
> Okay, I understand the purpose now, 100% agree it's useful. I also think your last criticism is absolutely valid, though I'm less on board with your proposed solution. I think I'd prefer an array field or even global field for each class, something like that. I don't like this idea of infinitely hooking this function to add a single item to the array.

Yeah that was my original idea too but then @TaroEld proposed adding a function for that so that people have the possibility of modifying the stats for exotic weapons or if they want for niche cases - beyond just copying from the base item.

LordMidas commented:
Regarding using a global array for `getFieldsForRandomize` there is a slight "inconvenience" issue in that the fields are different for each named item type. So currently we'd need 3 separate global arrays for each of the following: named_weapon, named_shield, named_armor. If we just implement the function without a global array then each of these classes can have the list defined within that own custom definition of the function. So, with that in mind, what's the final word? Global arrays or keep inside function?

Enduriel commented:
please target #316 instead of development

Enduriel reviewed (commented):
Thoughts on creating a separate function to call inside of each of these which would apply the shared parts of the hooks? Specifically:
```
q.m.BaseItemScript <- null;
q.setValuesBeforeRandomize <- function( _values )
q.randomizeValues = @(__original) function()
```
?

LordMidas committed:
refactor: use getFieldsForRandomize

LordMidas committed:
feat: add namespace to add helper functions for hooks

LordMidas committed:
refactor: use helper function to add shared hooks to named items

LordMidas committed:
refactor: use verylate bucket for wrapping instead of hookTree

LordMidas committed:
refactor: delete HooksHelper table at the end of load

LordMidas committed:
fix: use equals instead of new slot for wrapping hook

LordMidas committed:
refactor: remove setValuesBeforeRandomize and bake into randomizeValues

LordMidas committed:
refactor: change helper function to use a single function

TaroEld commented on msu/load.nut:
```diff
@@ -7,7 +7,10 @@ local function includeLoad( _folder )
 
 includeLoad("ui");
 includeLoad("systems");
+includeLoad("hooks_helper");
```
Not sure why we have a folder and then a file that's named the same and include both ish

TaroEld replied:
But ok I guess you want to put it into different files later or sth.

Enduriel replied:
yeah we might have more stuff I suspect, not a huge fan of how it's organized rn but hey that's for later us to fix when it becomes relevant I think.

TaroEld commented on msu/hooks/items/weapons/named/named_weapon/named_weapon.nut:
```diff
@@ -0,0 +1,22 @@
+			"ConditionMax",
+			"StaminaModifier",
+			"RegularDamage",
+			"RegularDamageMax"
+			"ArmorDamageMult"
+			"ChanceToHitHead"
+			"DirectDamageMult"
+			"DirectDamageAdd"
+			"StaminaModifier"
+			"ShieldDamage"
+			"AdditionalAccuracy"
+			"FatigueOnSkillUse"
```
This would appear to be missing commas

Enduriel committed:
fix: missing commas in array definition

LordMidas committed:
refactor: add setValuesBeforeRandomize function

Having a function allows people to hook it to add logic e.g. by extracting values from base item via function calls instead of only m table.

LordMidas committed:
refactor: change getFieldsForRandomize function to BaseItemFields array

TaroEld commented on msu/hooks/items/armor/named/named_armor.nut:
```diff
@@ -0,0 +1,9 @@
+::MSU.HooksHelper.addBaseItemToNamedItem("scripts/items/armor/named/named_armor");
+
+::MSU.HooksMod.hook("scripts/items/armor/named/named_armor", function(q) {
+	q.m.BaseItemFields = [
+		"Condition",
+		"ConditionMax",
+		"StaminaModifier"
+	];
+});
```
Might as well pass the array into your helper and set it there.

LordMidas committed:
fix: use binary operator to add ItemType instead of overwriting it

Because named items have the `Named` ItemType and we don't want to overwrite that but only add to that.

LordMidas committed:
refactor: change from array to function to get base item fields

LordMidas committed:
fix: add more fields to be copied from base item where relevant

Enduriel reviewed (approved):
This PR is pretty much a lesson in why composition >>> inheritance. Generally lgtm in this state.

LordMidas committed:
refactor: change to hookTree on create to set values before randomize

This ensures that the values from the base item are correctly copied even in the case someone overwrites randomizeValues.

LordMidas committed:
refactor: use a switcheroo instead of boolean field for randomizeValues

LordMidas committed:
fix: replace HooksMod with MH

LordMidas committed:
refactor: delete hookshelper at the end of FirstWorldInit bucket

LordMidas committed:
refactor: pass an instance of base item to setValuesBeforeRandomize

LordMidas committed:
fix: manually re-add Named itemtype instead of handling it differently

We re-add it directly in our hook on create because if someone overwrites setValuesBeforeRandomize we still want to ensure that the Named ItemType gets added back. We add in a check for isItemType for Named because some mod may want to remove it from certain items (in theory) so this is robust against that.

LordMidas committed:
refactor: move setting of Named itemtype to setValuesBeforeRandomize

If someone is overwriting this function then its their job to make sure the Named itemtype is added back correctly.

Enduriel commented on msu/hooks_helper/hooks_helper.nut:
```diff
@@ -0,0 +1,43 @@
+						this.m[field] = _baseItem.m[field];
+					}
+
+					this.m.ItemType = this.m.ItemType | ::Const.Items.ItemType.Named; // Re-add the named item type that got overwritten from the base item
```
Should be moved into a separate hookTree

TaroEld commented on msu/load.nut:
```diff
@@ -7,6 +7,7 @@ local function includeLoad( _folder )
 
 includeLoad("ui");
 includeLoad("systems");
+includeLoad("hooks_helper");
```
We discussed and we should probably expose this for when a mod adds another named item type?

LordMidas replied:
Yes that sounds good to me.

LordMidas commented:
Postponed as I am going to first implement, test and polish this system over at the Modular Vanilla mod.
